File: Stack/Easy/valid_substring.cpp
// https://practice.geeksforgeeks.org/problems/valid-substring0624/1#
// { Driver Code Starts
// Initial template for C++
#include <bits/stdc++.h>
using namespace std;


 // } Driver Code Ends
// User function template for c++
class Solution {
  public:
    int findMaxLen(string s) {
        stack<int> st;
        int n = s.size();
        int ans = 0;
        st.push(-1);
        for(int i = 0; i < n; i++)
        {
            if(s[i] == '(') st.push(i);
            else
            {
                st.pop();
                
                if(st.empty()) st.push(i);
                else
                {
                    ans = max(ans, i- st.top());
                }
            }
        }
        
        // ans = max(ans, curr);
        return ans;
    }
};

// (()))())
// 

// { Driver Code Starts.
int main() {
    int t;
    cin >> t;
    while (t--) {
        string S;
        cin >> S; 
        Solution ob;
        cout << ob.findMaxLen(S) << endl;
    }
    return 0;
}
  // } Driver Code Ends

File: Stack/Easy/valid_substring.py
# https://practice.geeksforgeeks.org/problems/valid-substring0624/1#
#User function Template for python3
# https://practice.geeksforgeeks.org/problems/valid-substring0624/1#
class Solution:
    def findMaxLen(ob, S):
        st = []
        ans = 0
        st.append(-1)
        
        for i in range(len(S)):
            if(S[i] == '('):
                st.append(i)
            else:
                st.pop()
                if(len(st)>0):
                    ans = max(ans , i - st[-1]);
                else:
                    st.append(i)
        
        return ans;


#{ 
#  Driver Code Starts
#Initial Template for Python 3

if __name__ == '__main__': 
    t = int (input ())
    for _ in range (t):
        S=input()
        
        ob = Solution()
        print(ob.findMaxLen(S))
# } Driver Code Ends